### PR TITLE
Reader: add high watermark redux bits

### PR DESCRIPTION
**Pull request by samouri**
This PR adds a relatively simple action/reducer/schema for handling high watermarks.

The shape of the reduced is roughly:
```js
{
 [ stream_key ]: date numeral,
}
```

The reducer will listen for actions of type `READER_VIEW_STREAM` and will raise the watermark for that stream assuming it is newer than what was previously stored.

None of this is actually live code yet, so it won't be included in our bundles.

open question
----
what should we use as a date format? I've picked integers for now, but i'm open to changing it (ideally to strings).

1. a string standard would be the most flexible and serializes well (keeping in mind timezones).  whats bad about strings is that we will be constantly converting between it and integers for comparisons.
2. moment would be too heavy IMO
3. numbers.  we can do tz conversions when we receive from the api. only weirdness will be if someone changes their laptops tz or goes on a flight.

To test
----
- unit test coverage.  `npm run test-client`. or trust CI :)

### CODE REVIEW

**Comment by blowery** [comment]
> what should we use as a date format?

HWM isn't strictly a date. In some cases, we may need more info that just the date to break ties.

So I vote string. We'll just have to make sure the marks are string comparable. like `'YYYYMMDDTHHmmSS-{paddedPostId}'` or something.

**Comment by blowery** [comment]
also, TZ shouldn't really enter into anything. Any dates we serialize we should serialize as ISO, not at a specific TZ.

**Comment by samouri** [comment]
potentially dumb question: are ties real? and assuming they are, what is the issue with them?  You've either read both or haven't read either

**Comment by blowery** [comment]
Yeah, ties based on date are totally real and weirdly common. You'd be amazed how many sites publish content on the hour. 

I'm unsure if they're an issue here or not. I'm leaning towards not. They _are_ an issue when trying to page through the stream (imagine you have two pages of posts all at the same time), but for HWM... Not entirely sure I care as much.

**Comment by blowery** [comment]
OK, let's start with a "Date" (timestamp from `Date.now()` should work).

File: client/state/reader/watermarks/test/reducer.js [hunk 1]
```diff
@@ -0,0 +1,41 @@
+/** @format */
+
+/**
+ * Internal dependencies
+ */
+import { viewStream } from '../actions';
+import { watermarks } from '../reducer';
+
+const streamId = 'special-chicken-stream';
```
Comment (bluefuton) [inline client/state/reader/watermarks/test/reducer.js]: 🐔

File: client/state/reader/watermarks/watermark-schema.js [hunk 2]
```diff
@@ -0,0 +1,10 @@
+/** @format */
+
+export default {
+	type: 'object',
+	additionalProperties: false,
+	patternProperties: {
+		// todo: does this match everything? thats what i want
```
Comment (bluefuton) [inline client/state/reader/watermarks/watermark-schema.js]: Can we remove this todo?

File: client/state/reader/watermarks/test/reducer.js (hunk 1 above)
Comment (blowery) [reply client/state/reader/watermarks/test/reducer.js]: 🥇

File: client/state/reader/watermarks/watermark-schema.js (hunk 2 above)
Comment (samouri) [reply client/state/reader/watermarks/watermark-schema.js]: i'm happy i added the todo, i still need to verify it

Comment (blowery) [reply client/state/reader/watermarks/watermark-schema.js]: i think you can just leave properties and patternProperties off entirely and set `additonalProperties: { type: 'number' }`?

Comment (blowery) [reply client/state/reader/watermarks/watermark-schema.js]: either way, we should have tests for the schema.

**Comment by samouri** [comment]
i'll add a schema test

File: client/state/reader/watermarks/watermark-schema.js (hunk 2 above)
Comment (samouri) [reply client/state/reader/watermarks/watermark-schema.js]: adding tests for deserialization now. that will test the schema

**Comment by samouri** [comment]
added tests for the schema in https://github.com/Automattic/wp-calypso/pull/18769/commits/cee4eb0f73e2cd6172a8fe3268a565b5493c14f9

**Comment by bluefuton** [comment]
Schema tests look good to me 👍

**Pull request merged**